src/metaopt/algo/space.py

```diff
@@ -0,0 +1,536 @@
+           RandomState instance **just for the call of this function**.
+           Default is None.
+
+           Set random state to something other than None for reproducible
+           results.
+
+        .. warning:: Setting `seed` with an integer will cause the same ndarray
+           to be sampled if ``n_samples > 0``. Set `seed` with a
+           ``numpy.random.RandomState`` to carry on the changes in random state
+           across many samples.
+
+        """
+        samples = [self.prior.rvs(*self._args, size=self._shape,
+                                  random_state=seed,
+                                  **self._kwargs) for _ in range(n_samples)]
+        return samples
+
+    def interval(self, alpha=1.0):
+        """Return a tuple containing lower and upper bound for parameters.
+
+        If parameters are drawn from an 'open' supported random variable,
+        then it will be attempted to calculate the interval from which
+        a variable is `alpha`-likely to be drawn from.
+
+        .. note:: Lower bound is inclusive, upper bound is exclusive.
+
+        """
+        return self.prior.interval(alpha, *self._args, **self._kwargs)
+
+    def __contains__(self, point):
+        """Check if constraints hold for this `point` of `Dimension`.
+
+        :param x: a parameter corresponding to this `Dimension`.
+        :type x: numeric or array-like
+
+        .. note:: Default `Dimension` does not have any extra constraints.
+           It just checks whether point lies inside the support and the shape.
+
+        """
+        low, high = self.interval()
+        point_ = numpy.asarray(point)
+        if point_.shape != self.shape:
+            return False
+        return numpy.all(point_ < high) and numpy.all(point_ >= low)
+
+    def __repr__(self):
+        """Represent the object as a string."""
+        return "{0}(name={1}, prior={{{2}: {3}, {4}}}, shape={5})".format(
+            self.__class__.__name__, self.name, self._prior_name,
+            self._args, self._kwargs, self.shape)
+
+    @property
+    def name(self):
+        """See `Dimension` attributes."""
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        if isinstance(value, str) or value is None:
+            self._name = value
+        else:
+            raise TypeError("Dimension's name must be either string or None. "
+                            "Provided: {}, of type: {}".format(value, type(value)))
+
+    @property
+    def type(self):
+        """See `Dimension` attributes."""
+        return self.__class__.__name__.lower()
+
+    @property
+    def shape(self):
+        """Return the shape of dimension."""
+        _, _, _, size = self.prior._parse_args_rvs(*self._args,  # pylint:disable=protected-access
+                                                   size=self._shape,
+                                                   **self._kwargs)
+        return size
+
+
+class _Discrete(Dimension):
```

Comment:
I believe _Discrete class is useless. The `sample` method could solely reside inside `Integer` class. `Categorical`'s `sample` method is straightforward using numpy.random.choice. Also, `interval` method is overwritten anyway by Categorical's to raise RuntimeError.

Although I could easily see how `Integer` and `Categorical` are related in theory, in our context combining them is not giving any advantage. It complexifies Categorical without giving any advantage in return.

Reply:
:sweat:

Reply:
I am not able to pass somehow a separate "sandboxed" seed without affecting global numpy RNG. `numpy.random.choice` does not offer such a choice. Temp saving the random state by:

```python
orig_state = numpy.random.get_state()
numpy.random.seed(seed)
samples = # Do stuff with `numpy.random` here
numpy.random.set_state(orig_state)
return samples
```

i am not sure whether affecting global numpy RNG is safe, if this is not your intention.

Reply:
What about setting a `self.rng = numpy.RandomState()` object inside the `Dimension`? We could then use `rng.choices()`.

Reply:
I will use it only inside `Categorical`. The rest of `Dimension` use `scipy.stats` which does this thing internally.

Reply:
I patched Categorical to use `numpy.RandomState`, using `check_random_state` helper function.

However, I found a difficulty removing the `_Discrete` subclass. Remember that `_Discrete` class was introduced as an alternative to remove `discrete` keyword argument from `Dimension`. So the programming flow currently is as follows, and it takes advantage of Python mechanics regarding class inheritance:

**Currently**:
Suppose a call to `Integer(...).sample(...)`. In general calling a method in an `Integer` object will first search within `Integer` and then `super` calls from there will try to find corresponding methods first in `Real`, then `_Discrete` and then `Dimension`.

```python
class Integer(Real, _Discrete):
    def interval(self, alpha=1.0):
        low, high = super(Integer, self).interval(alpha)  # Call to Real.interval
        # make them integers
        return int_low, int_high

    def __contains__(self, point):
        # Check if they are integers...
        return super(Integer, self).__contains__(point)  # Call to Dimension.__contains__


    # calling Integer(...).sample(...) will enter first Real.sample;
    # due to the order of subclasses
```
```python
class Real(Dimension):
    def interval(self, alpha=1.0):
        prior_low, prior_high = super(Real, self).interval(alpha)  # Call to Dimension.interval
        return (max(prior_low, self._low), min(prior_high, self._high))  # Add constraints to interval  

    def sample(self, n_samples=1, seed=None):
        # Checks if constraint are met!
        samples = []
        for _ in range(n_samples):
            for _ in range(4):
                sample = super(Real, self).sample(1, seed)  # Calls _Discrete.sample
                if sample[0] not in self:
                # Calls Integer.__contains__
                    # ...
            # ...
        # ...

        return samples
```

```python
class _Discrete(Dimension):
    def sample(self, n_samples=1, seed=None):
        samples = super(_Discrete, self).sample(n_samples, seed)  # Calls Dimension.sample
        # Making discrete by ourselves because scipy does not use **floor**
        return list(map(lambda x: numpy.floor(x).astype(int), samples))  # Converts to integers
```

So the sequence of calls goes like this:
* [`Integer.sample` (redirects)]
* `Real.sample` (fulfills extra constraints)
   * `_Discrete.sample` (converts to integers)
      * `Dimension.sample`  (samples from distribution)
   * `Integer.__contains__` (checks if they are integers)
      * `Dimension.__contains__` (check if they are within bounds)
         * `Integer.interval` (makes it int)
           * `Real.interval`  (extra constraints)
             * `Dimension.interval` (asks the distribution for its support)

Reply:
If we remove `_Discrete` and move all methods in `class Integer(Real)`. The flow will be become:

So the sequence of calls goes like this:
* `Integer.sample` (converts to integers)
   * `Real.sample` (fulfills extra constraints)
      * `Dimension.sample`  (samples from distribution)
      * `Integer.__contains__` (checks if they are integers)
         * `Dimension.__contains__` (check if they are within bounds)
            * `Integer.interval` (makes it int)
              * `Real.interval`  (extra constraints)
                * `Dimension.interval` (asks the distribution for its support)

Poor `Dimension.sample` will return whatever the scipy distribution returns, and the check from the `Integer.__contains__` will fail, if the underlying distribution in scipy corresponds to a real one... However, we should allow (I believe) discretizing real distributions, so that for instance someone could have approximately normal distributed integer stochastic variables. What do you think about this?

Reply:
I agree with the idea of discretizing Real distributions. However I don't see the problem here, can't we just move `_Discrete.sample()` inside `Integer.sample()` rather than completely removing this method and rely solely on `Real.sample()`?

Reply:
Sorry for the confusion. I meant if we removed `class _Discrete(Dimension)` and moved all of its methods to `class Integer(Real)` (including `sample`), then:

... the sequence of calls goes like this:

* `Integer.sample` (converts to integers)
   * `Real.sample` (fulfills extra constraints)
      * `Dimension.sample` (samples from distribution)  **-> this will return possibly samples in real numbers**
      * `Integer.__contains__` (checks if they are integers) **-> it will fail because it checks if they are integers**
         * `Dimension.__contains__` (check if they are within bounds)
           * `Integer.interval` (makes it int)
              * `Real.interval` (extra constraints)
                 * `Dimension.interval` (asks the distribution for its support)

Reply:
Oh yes indeed, I was confused. :sweat_smile: 

My proposition then would be to replace this confusing inheritance path with a simple separation of `sample()` and `_sample()` where sample makes verifications over `_sample()` which do it blindly.

So this line:
```python
sample = super(Real, self).sample(1, seed)
```
becomes 
```python
sample = self._sample(1, seed)
```

where

```python
class Real(Dimension):
    # ...
    def _sample(self, n_samples, seed=None):
        return super(Real, self).sample(1, seed)
    # ...

class Integer(Dimension):
    # ...
    def _sample(self, n_samples, seed=None):
        samples = super(Integer, self)._sample(n_samples, seed)
        return list(map(lambda x: numpy.floor(x).astype(int), samples))
    # ...
```

I think it is easier this way to track the path of execution in order to understand, modify and maintain the code.

We would now have:

* `Integer.sample` (converts to integers)
   * `Real.sample` (fulfills extra constraints)
      * `Integer._sample`
          * `Real._sample`
             * `Dimension._sample` (samples from distribution)
      * `Integer.__contains__` 
         * `Dimension.__contains__` (check if they are within bounds)
           * `Integer.interval` (makes it int)
              * `Real.interval` (extra constraints)
                 * `Dimension.interval` (asks the distribution for its support)